### ISSUE TITLE
Randomized transforms applied differently to image and segmentation  in on Dataset

### ISSUE THREAD

**Issue by relyativist** [Project-MONAI/MONAI#4545]
**Describe the bug**
As follows from title, I am trying to create `monai.data.Dataset` and apply transform via `monai.transforms.Compose`, each time I call` train_loader.dataset` - I have different transformation for image and segmentation. 

**To Reproduce**
1) Create instance of `monai.data.Dataset() and monai.data.DataLoader()` with
`data =  {'image': ['/tmp/sub-1/sub-1_t1.nii.gz', '/tmp/sub-t2.nii.gz'], 'seg': '/tmp/1.nii.gz'}`
 and transform equal to

```
train_transf = Compose(
    [
        LoadImaged(keys=["image", "seg"]),
        EnsureChannelFirstd(keys=["image", "seg"]),
        RandRotated(keys=["image", "seg"], range_x=0.0, range_y=0.0, range_z=0.75, prob=1),
        Resized(keys=["image", "seg"], spatial_size=(128,128,128), mode=('area')),
        EnsureTyped(keys=["image", "seg"], dtype=torch.float),
    ]
)

train_ds = monai.data.Dataset(data=train_files, transform=train_transf)
train_loader = DataLoader(
    train_ds,
    batch_size=2,
    shuffle=False,
    num_workers=0,
    collate_fn=list_data_collate,
    pin_memory=torch.cuda.is_available(),
)

``` 
2) Call train_loader.dataset for dataset index multiple times to ensure image and segmentation rotate together:

```
ind = 25
for i in range(10):
    image, label = train_loader.dataset[ind]["image"], train_loader.dataset[ind]["seg"]
    lab_loc = round(np.where(label>0)[3].shape[0] / 2) # find where label values are not null to plot as overlay over image
    lab_pos = np.where(label>0)[3][lab_loc]
    plt.figure("visualise", (8, 4))
    plt.subplot(1, 2, 1)
    plt.title("image")
    plt.imshow(image[0,:, :, lab_pos], cmap='gray')
    #plt.subplot(1, 2, 2)
    plt.title("label")
    plt.imshow(label[0,:, :, lab_pos], cmap="Reds", alpha=0.2)
    plt.show()

```

**Expected behavior**
I expect to see that image and segmentation rotate constantly with same direction, I attach the link to image, to show ground truth segmentation position and second one, what I get after transformation.

**Screenshots**
[Ground truth label](https://disk.yandex.com/i/qrzptZsMqDXZeQ)
[plot on train_loader.dataset call with transformation](https://disk.yandex.com/i/2bY_HN3sEnlFPQ)

**Environment**

Ensuring you use the relevant python executable, please paste the output of:

```
================================
Printing MONAI config...
================================
MONAI version: 0.9.0
Numpy version: 1.22.4
Pytorch version: 1.8.1+cu102
MONAI flags: HAS_EXT = False, USE_COMPILED = False
MONAI rev id: af0e0e9f757558d144b655c63afcea3a4e0a06f5
MONAI __file__: /root/miniconda3/lib/python3.8/site-packages/monai/__init__.py

Optional dependencies:
Pytorch Ignite version: NOT INSTALLED or UNKNOWN VERSION.
Nibabel version: 3.2.0
scikit-image version: 0.19.3
Pillow version: 9.1.0
Tensorboard version: 2.9.0
gdown version: NOT INSTALLED or UNKNOWN VERSION.
TorchVision version: NOT INSTALLED or UNKNOWN VERSION.
tqdm version: 4.64.0
lmdb version: NOT INSTALLED or UNKNOWN VERSION.
psutil version: 5.9.0
pandas version: 1.4.2
einops version: 0.4.1
transformers version: NOT INSTALLED or UNKNOWN VERSION.
mlflow version: NOT INSTALLED or UNKNOWN VERSION.
pynrrd version: NOT INSTALLED or UNKNOWN VERSION.

For details about installing the optional dependencies, please visit:
    https://docs.monai.io/en/latest/installation.html#installing-the-recommended-dependencies


================================
Printing system config...
================================
System: Linux
Linux version: Ubuntu 20.04.3 LTS
Platform: Linux-5.4.0-113-generic-x86_64-with-glibc2.17
Processor: x86_64
Machine: x86_64
Python version: 3.8.12
Process name: python
Command: ['python', '-c', 'import monai; monai.config.print_debug_info()']
Open files: []
Num physical CPUs: 20
Num logical CPUs: 40
Num usable CPUs: 28
CPU usage (%): [0.5, 0.5, 1.0, 0.5, 5.8, 4.3, 1.9, 2.9, 1.0, 0.5, 1.4, 0.0, 6.2, 100.0, 100.0, 6.3, 7.2, 100.0, 100.0, 6.3, 6.7, 100.0, 15.9, 100.0, 10.0, 6.3, 7.2, 6.3, 99.5, 6.7, 11.0, 6.7, 6.7, 6.2, 3.4, 6.7, 6.2, 3.4, 2.9, 12.1]
CPU freq. (MHz): 2
Load avg. in last 1, 5, 15 mins (%): [15.8, 16.2, 16.3]
Disk usage (%): 90.8
Avg. sensor temp. (Celsius): UNKNOWN for given OS
Total physical memory (GB): 251.6
Available memory (GB): 237.9
Used memory (GB): 11.6

================================
Printing GPU config...
================================
Num GPUs: 4
Has CUDA: True
CUDA version: 10.2
cuDNN enabled: True
cuDNN version: 7605
Current device: 0
Library compiled for CUDA architectures: ['sm_37', 'sm_50', 'sm_60', 'sm_70']
GPU 0 Name: NVIDIA GeForce GTX 1080 Ti
GPU 0 Is integrated: False
GPU 0 Is multi GPU board: False
GPU 0 Multi processor count: 28
GPU 0 Total memory (GB): 10.9
GPU 0 CUDA capability (maj.min): 6.1
GPU 1 Name: NVIDIA GeForce GTX 1080 Ti
GPU 1 Is integrated: False
GPU 1 Is multi GPU board: False
GPU 1 Multi processor count: 28
GPU 1 Total memory (GB): 10.9
GPU 1 CUDA capability (maj.min): 6.1
GPU 2 Name: NVIDIA GeForce GTX 1080 Ti
GPU 2 Is integrated: False
GPU 2 Is multi GPU board: False
GPU 2 Multi processor count: 28
GPU 2 Total memory (GB): 10.9
GPU 2 CUDA capability (maj.min): 6.1
GPU 3 Name: NVIDIA GeForce GTX 1080 Ti
GPU 3 Is integrated: False
GPU 3 Is multi GPU board: False
GPU 3 Multi processor count: 28
GPU 3 Total memory (GB): 10.9
GPU 3 CUDA capability (maj.min): 6.1
```

**Additional context**

**Comment by wyli** [comment]
thanks for reporting, I think that's the same question as this https://github.com/Project-MONAI/MONAI/issues/1063 each get_item from a dataset will generate a randomized outcome.

you can use:
```py
item = train_loader.dataset[ind]
image, label = item["image"], item["seg"]
```

**Comment by relyativist** [comment]
@wyli Many thanks for Answer, however I don't understand how then `item` should be passed to training loop as iterable? As an input to `one_epoch` function?
```
def one_epoch(model, 
                criterion, 
                opt, 
                config, 
                dataloader, 
                device, 
                writer, 
                epoch, 
                metric_dict_epoch, 
                n_iters_total=0,
                augmentation=None, 
                is_train=True):


    # use amp to accelerate training
    if config.opt.use_scaler:
        scaler = torch.cuda.amp.GradScaler()

    phase_name = 'train' if is_train else 'val'
    loss_name = config.opt.criterion
    metric_dict = defaultdict(list)
    target_metric_name = config.model.target_metric_name 

    if not is_train:
        model.eval()
    else:
        model.train()

    # used to turn on/off gradients
    grad_context = torch.autograd.enable_grad if is_train else torch.no_grad
    with grad_context():
        iterator = enumerate(dataloader)
        val_predictions = {}
        for iter_i, data_tensors in iterator:
            brain_tensor = data_tensors['image']
            label_tensor = data_tensors['seg']
```

I still get different transformations for image and segmentation.

**Comment by wyli** [comment]
The way you use it looks good to me, but it's difficult to be sure without a fully reproducible example. There are some training pipelines from the tutorials, please refer to https://github.com/Project-MONAI/tutorials/tree/main/3d_segmentation
